### DISCUSSION TITLE
Create multiple prisma clients dynamically for same schema

### DISCUSSION THREAD

**Discussion by andreasvh-conceto** [prisma/prisma#17714] [Q&A] [answered]
Hello,
currently we have a backend with one prisma client, where an environment variable holds the database url. Currently every customer gets its own database with its own backend , an own backend app instance with its own prisma client and  a single database_url environment variable - a classic  default scenario.

Now the idea is to create just one backend, where I can instantiate  multiple prisma client instances on startup. The schema is for all the same. They are just connected to different databases. Each customer should get its own prisma Client, where each prisma client should be connected to the related customer database. Each connection url has a name (customername) and its related url. This could be stored in an environment variable $DATABASE_URLS, which hold a json array  of objects like [ {name: "customer1" ,url: "databaseUrlCustomer1?pgbouncer=true"}, ...]  - whatever. On startup for each database url an prisma client instance should be created for each of these url objects. In the backend code I want to receive the right prisma instance with a service which holds a key value pair, where key is the customer name and the value the related prisma instance. 

Is it possible to implement such a scenario? Could you give me a code example how to instantiate a prisma Client  like  new PrismaClient(name,url) or something like that?  Is there something I need to consider in prisma.schema file?

Best regards and thank you for your help
Andreas

**Comment by nurul3101** [comment, marked as answer]
Hello @andreasvh-conceto 👋 

What you are trying to achieve is possible, there are ways through which you can achieve this. I would recommend you to go through this Feature Request:
- https://github.com/prisma/prisma/issues/2443

And specifically this [comment](https://github.com/prisma/prisma/issues/2443#issuecomment-630679118) that discusses how you can have multiple PrismaClient instances connecting to different databases.

Can you have a look at it?

**Reply by andreasvh-conceto** [reply]
Thank you that was exactly what i was looking for. This comment helped me finally out:

 https://github.com/prisma/prisma/issues/2443#issuecomment-630679118

**Comment by michael-land** [comment]
Does creating multiple  (hundreds) Prisma clients significantly impact memory usage or performance? In a multi-tenant model with one database per tenant, where a serverless function might handle many requests and initialize hundreds of Prisma clients, could this lead to performance degradation?

**Reply by andreasvh-conceto** [reply]
Hi, 
i can recommend this section in the docs regarding to your question: [connection pool size](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections#recommended-connection-pool-size)

For multi tenancy approach you have to set the `connection_limit` parameter for each prisma client.
Since 1,5 years we are running a multi tenancy application successfully and have a 16 core CPU. Currently we set the connection limit to 2 for small tenants (customers) and 3 for bigger ones and it works still fine, but we have a small user base. Currently we do not plan to go for more than 10 tenants per multi tenancy backend. In Azure cloud we configured  [pgbouncer](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections/pgbouncer#set-pgbouncer-to-transaction-mode) for deleting dead connections.

Related to serverless environments:
https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections#serverless-environments-faas
https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections#without-an-external-connection-pooler
https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections#recommended-connection-pool-size

Hope that helps :)
